resolwe/flow/models/utils.py

```diff
@@ -747,3 +747,68 @@ def referenced_files(data, include_descriptor=True):
     )
 
     return special_files + output_files + descriptor_files
+
+
+def get_collection_of_input_entities(data):
+    """Get collection that contains all "entity inputs" of a given data.
+
+    With "entity input", one refers to the inputs that are part of an entity.
+    """
+    # Prevent circular imports:
+    from resolwe.flow.models import Data, Collection
+
+    collections = set()
+
+    for field_schema, fields in iterate_fields(data.input, data.process.input_schema):
+        name = field_schema["name"]
+        value = fields[name]
+        if "type" not in field_schema:
+            continue
+        if field_schema["type"].startswith("data:"):
+            if value is None:
+                continue
+            try:
+                data = Data.objects.get(id=value)
```

Comment:
This is inefficient as a separate db query is made for each input and then an additional request is made for `data.collection`. You just need to collect data ids here and use them in a single request at the end of the function:
```python
collections = Collection.objects.filter(data__in=[data_ids])
if collections.count() != 1:
    raise

return collections.first()
```

Reply:
Not totally true, since we check that for each data it also holds that `data.entity != None`. But you're right that this can be optimized. Will do.